### PR TITLE
Kovan WASM fork code

### PR DESCRIPTION
**Pull request by NikVolf**
in chainspec json files, there is a breaking change with wasm flag, one should replace

`wasm: true` with `wasmActivationTransition: "0x0"`

### CODE REVIEW

File: ethcore/src/executive.rs [hunk 1]
```diff
@@ -152,8 +153,13 @@ impl TransactOptions<trace::NoopTracer, trace::NoopVMTracer> {
...
 
-pub fn executor(machine: &Machine, vm_factory: &Factory, params: &ActionParams) -> Box<vm::Vm> {
-	if machine.supports_wasm() && params.code.as_ref().map_or(false, |code| code.len() > 4 && &code[0..4] == WASM_MAGIC_NUMBER) {
+pub fn executor(
+	machine: &Machine,
+	vm_factory: &Factory,
+	params: &ActionParams,
+	blocknumber: BlockNumber,
+) -> Box<vm::Vm> {
+	if machine.supports_wasm(blocknumber) && params.code.as_ref().map_or(false, |code| code.len() > 4 && &code[0..4] == WASM_MAGIC_NUMBER) {
```
Comment (debris) [inline ethcore/src/executive.rs]: I don't like this change at all. The `if` statement that we have right now is terrible and this change makes it even worse. This function/interface needs to be completely reworked

File: ethcore/Cargo.toml [hunk 2]
```diff
@@ -70,6 +70,9 @@ unexpected = { path = "../util/unexpected" }
 journaldb = { path = "../util/journaldb" }
 tempdir = "0.3"
 
+[dev-dependencies]
+wabt = "0.1"
```
Comment (debris) [inline ethcore/Cargo.toml]: I refuse pulling a huge c++ dependency to ethcore as a mandatory dependency. It needs to be do differently.

File: ethcore/res/ethereum/kovan_wasm_test.json [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+		"forkBlock": 4297256,
+		"forkCanonHash": "0x0a66d93c2f727dca618fabaf70c39b37018c73d78b939d8b11efbbd09034778f",
+		"validateReceiptsTransition" : 1000000,
+		"eip155Transition": 1000000,
+		"validateChainIdTransition": 1000000,
+		"eip140Transition": 5067000,
+		"eip211Transition": 5067000,
+		"eip214Transition": 5067000,
+		"eip658Transition": 5067000,
+		"wasmActivation": 10
```
Comment (tomusdrw) [inline ethcore/res/ethereum/kovan_wasm_test.json]: The convention seems to be `*Transition`, would keep it.

File: ethcore/src/machine.rs [hunk 4]
```diff
@@ -377,8 +377,8 @@ impl EthereumMachine {
 	}
 
 	/// If this machine supports wasm.
-	pub fn supports_wasm(&self) -> bool {
-		self.params().wasm
+	pub fn supports_wasm(&self, blocknumber: BlockNumber) -> bool {
```
Comment (tomusdrw) [inline ethcore/src/machine.rs]: Do we need a separate public method for that? perhaps use `Schedule` directly on the call site.

File: ethcore/src/spec/spec.rs [hunk 5]
```diff
@@ -111,6 +111,8 @@ pub struct CommonParams {
 	pub remove_dust_contracts: bool,
 	/// Wasm support
 	pub wasm: bool,
```
Comment (tomusdrw) [inline ethcore/src/spec/spec.rs]: Why is that needed? Seems that it's just `wasmTransition: 0`. We already did breaking changes to spec files in the past, we should just highlight that in the release notes.

File: ethcore/vm/src/action_params.rs [hunk 6]
```diff
@@ -112,6 +112,14 @@ impl Default for ActionParams {
 	}
 }
 
+impl ActionParams {
+	pub fn from_origin(mut self, address: Address) -> Self {
```
Comment (tomusdrw) [inline ethcore/vm/src/action_params.rs]: Seems to be used only for tests, might be misleading to include it everwhere in the code.

File: ethcore/vm/src/schedule.rs [hunk 7]
```diff
@@ -113,6 +113,8 @@ pub struct Schedule {
 	pub kill_dust: CleanDustMode,
 	/// Enable EIP-86 rules
 	pub eip86: bool,
+	/// Is wasm active
+	pub wasm_activated: bool,
```
Comment (tomusdrw) [inline ethcore/vm/src/schedule.rs]: Maybe `wasm: Option<WasmCosts>` will be more future proof?

File: ethcore/src/executive.rs (hunk 1 above)
Comment (tomusdrw) [reply ethcore/src/executive.rs]: Yeah, perhaps it should just go to `vm_factory`? This function seems to be another factory that delegates to `vm_factory` in case it's not wasm.

```rust
vm_factory.create(params.gas, params.code, blocknumber)
```

File: ethcore/Cargo.toml (hunk 2 above)
Comment (NikVolf) [reply ethcore/Cargo.toml]: it's not mandatory, it's only for running tests

File: ethcore/src/executive.rs (hunk 1 above)
Comment (NikVolf) [reply ethcore/src/executive.rs]: I wanted to go this way, but last time I checked it was just not possible for some reason I can't recall. perhaps I should try again and find this reason again

File: ethcore/res/ethereum/kovan_wasm_test.json (hunk 3 above)
Comment (NikVolf) [reply ethcore/res/ethereum/kovan_wasm_test.json]: "wasmTransition" might imply that evm no longer works, i think it's not the best convention in this case

Comment (tomusdrw) [reply ethcore/res/ethereum/kovan_wasm_test.json]: Maybe `enableWasmTransition` or `supportWasmTransition` then?

Comment (NikVolf) [reply ethcore/res/ethereum/kovan_wasm_test.json]: both `enableWasmTransition` and `supportWasmTransition` implies the possibilty of such transition at all, not some blocknumber where it is activated

it is not really about transition as eip-s

Comment (tomusdrw) [reply ethcore/res/ethereum/kovan_wasm_test.json]: `supportWasmTransition` and `enableWasmTransition` implies for me that `wasm is supported` or `wasm is enabled` starting at some block. Similar to how `eip658Transition` implies that `eip658` is active at some point (it doesn't disable any previous eips).
That said my main reason is to complain is to keep the naming consistent here, if in the future we support a fork to wasm-only network we can either have: `wasmOnlyTransition` or `disableEvmTransition` there. Introducing another convention will make the file even messier imho.

File: ethcore/Cargo.toml (hunk 2 above)
Comment (debris) [reply ethcore/Cargo.toml]: It's mandatory for everyone working on daily basis on the project. It adds `cmake` with all of it's burden to our toolchain.

Comment (NikVolf) [reply ethcore/Cargo.toml]: np, but it will be just bytecode dump as a test

we already have c++ build tools prerequisites by the way (rocksdb will be first notable example), they just don't use cmake crate directly and instead use some obscure build scripts, so this cmake reference does not add anything to the list instead of one-time compilation of wabt

File: ethcore/vm/src/schedule.rs (hunk 7 above)
Comment (NikVolf) [reply ethcore/vm/src/schedule.rs]: how do you imagine this line then:

```
		if self.wasm || block_number >= self.wasm_activation {
			schedule.wasm_activated = true;
		}

```

in `CoimmonParams::update_schedule`

`schedule.wasm = Some(...<WHAT?>)` ?

where the prices will be held until they are suddenly active?

Comment (NikVolf) [reply ethcore/vm/src/schedule.rs]: On the second thought, no need to keep them, they can be just default :)

**Comment by NikVolf** [comment]
@debris , @tomusdrw 

all addressed!

File: ethcore/vm/src/schedule.rs (hunk 7 above)
Comment (tomusdrw) [reply ethcore/vm/src/schedule.rs]: Yeah, they can be default for now, but later probably come from spec (that's why I suggested merging them with the flag).

File: ethcore/wasm/src/lib.rs [hunk 8]
```diff
@@ -79,8 +79,8 @@ impl vm::Vm for WasmInterpreter {
 			&wasmi::ImportsBuilder::new().with_resolver("env", &instantiation_resolover)
 		).map_err(Error)?;
 
-		let adjusted_gas = params.gas * U256::from(ext.schedule().wasm.opcodes_div) /
-			U256::from(ext.schedule().wasm.opcodes_mul);
+		let adjusted_gas = params.gas * U256::from(ext.schedule().wasm().opcodes_div) /
```
Comment (tomusdrw) [inline ethcore/wasm/src/lib.rs]: Each call to `ext.schedule()` is virtual, can we cache it in a variable somewhere at the top?

File: ethcore/wasm/src/parser.rs [hunk 9]
```diff
@@ -24,12 +24,12 @@ use parity_wasm::peek_size;
 fn gas_rules(schedule: &vm::Schedule) -> rules::Set {
 	rules::Set::new({
 		let mut vals = ::std::collections::HashMap::with_capacity(4);
-		vals.insert(rules::InstructionType::Load, schedule.wasm.mem as u32);
-		vals.insert(rules::InstructionType::Store, schedule.wasm.mem as u32);
-		vals.insert(rules::InstructionType::Div, schedule.wasm.div as u32);
-		vals.insert(rules::InstructionType::Mul, schedule.wasm.mul as u32);
+		vals.insert(rules::InstructionType::Load, schedule.wasm().mem as u32);
```
Comment (tomusdrw) [inline ethcore/wasm/src/parser.rs]: Just wondering, maybe all these methods could just do with `WasmCost` instead of entire schedule?

File: ethcore/wasm/src/lib.rs (hunk 8 above)
Comment (NikVolf) [reply ethcore/wasm/src/lib.rs]: I don't think so, it's mutably borrows self all the way down the end of the function few times

I think it will be also optimized / devirtualized anyway

**Comment by rphmeier** [comment]
Prefer to put this in after #7605

File: ethcore/res/ethereum/kovan_wasm_test.json [hunk 10]
```diff
@@ -0,0 +1,74 @@
+{
+	"name": "Kovan",
+	"dataDir": "kovan",
```
Comment (rphmeier) [inline ethcore/res/ethereum/kovan_wasm_test.json]: these should reflect the test-ness of the spec

File: ethcore/res/ethereum/kovan_wasm_test.json [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+		"0x0000000000000000000000000000000000000005": { "builtin": { "name": "modexp", "activate_at": 5067000, "pricing": { "modexp": { "divisor": 20 } } } },
+		"0x0000000000000000000000000000000000000006": { "builtin": { "name": "alt_bn128_add", "activate_at": 5067000, "pricing": { "linear": { "base": 500, "word": 0 } } } },
+		"0x0000000000000000000000000000000000000007": { "builtin": { "name": "alt_bn128_mul", "activate_at": 5067000, "pricing": { "linear": { "base": 40000, "word": 0 } } } },
+		"0x0000000000000000000000000000000000000008": { "builtin": { "name": "alt_bn128_pairing", "activate_at": 5067000, "pricing": { "alt_bn128_pairing": { "base": 100000, "pair": 80000 } } } },
+		"0x00521965e7bd230323c423d96c657db5b79d099f": { "balance": "1606938044258990275541962092341162602522202993782792835301376" }
+	},
+	"nodes": [
+		"enode://56abaf065581a5985b8c5f4f88bd202526482761ba10be9bfdcd14846dd01f652ec33fde0f8c0fd1db19b59a4c04465681fcef50e11380ca88d25996191c52de@40.71.221.215:30303",
+		"enode://d07827483dc47b368eaf88454fb04b41b7452cf454e194e2bd4c14f98a3278fed5d819dbecd0d010407fc7688d941ee1e58d4f9c6354d3da3be92f55c17d7ce3@52.166.117.77:30303",
```
Comment (rphmeier) [inline ethcore/res/ethereum/kovan_wasm_test.json]: what are these bootnodes?

File: ethcore/vm/src/tests.rs [hunk 12]
```diff
@@ -85,6 +85,11 @@ impl FakeExt {
 		ext.schedule = Schedule::new_byzantium();
 		ext
 	}
+
+	pub fn with_wasm(mut self) -> Self {
```
Comment (rphmeier) [inline ethcore/vm/src/tests.rs]: doc-string

File: ethcore/res/ethereum/kovan_wasm_test.json (hunk 11 above)
Comment (NikVolf) [reply ethcore/res/ethereum/kovan_wasm_test.json]: it's just copy from `kovan.json`

**Comment by NikVolf** [comment]
@rphmeier it's easy to check (just merge 2 branches in 1), there is no conflicts with #7605 for now

File: json/src/spec/params.rs [hunk 13]
```diff
@@ -117,6 +115,11 @@ pub struct Params {
 	/// Transaction permission contract address.
 	#[serde(rename="transactionPermissionContract")]
 	pub transaction_permission_contract: Option<Address>,
+	/// Wasm support flag
+	pub wasm: Option<bool>,
```
Comment (tomusdrw) [inline json/src/spec/params.rs]: Is it still needed? What if it's set to `wasm: true && wasm_activiation_transition: 500` is it activated right away or at block `500`?

IMHO we should keep spec files clean (even if it requires a migration of existing spec files), especially given `wasm` was always experimental to date.

File: ethcore/src/factory.rs [hunk 14]
```diff
@@ -15,14 +15,44 @@
...
+
+/// Virtual machine factory
+#[derive(Default, Clone)]
+pub struct VmFactory {
+	evm: EvmFactory,
+}
+
+impl VmFactory {
+	pub fn create(&self, params: &ActionParams, schedule: &Schedule) -> Box<Vm> {
+		if schedule.wasm.is_some() && params.code.as_ref().map_or(false, |code| code.len() > 4 && &code[0..4] == WASM_MAGIC_NUMBER) {
```
Comment (tomusdrw) [inline ethcore/src/factory.rs]: Does WASM work with `gas > u64::max_value()`? I think I saw couple of `low_u64` calls below.

Comment (NikVolf) [reply ethcore/src/factory.rs]: It will fail later in the vm itself:

https://github.com/paritytech/parity/blob/master/ethcore/wasm/src/lib.rs#L85

File: json/src/spec/params.rs (hunk 13 above)
Comment (NikVolf) [reply json/src/spec/params.rs]: no, not needed, probably should mention that it requires to change in json spec to replace

`wasm: true` with `wasmActivationTransition: "0x0"`
